### PR TITLE
Query Builder

### PR DESCRIPTION
**Pull request by Renzenar**
Basic Foundation of the Query Builder

### CODE REVIEW

File: Team_Roster/immaculateGridCalculations/gridSolver.py [hunk 1]
```diff
@@ -1,6 +1,110 @@
...
+from sqlalchemy import func, and_
+
+
+def getAllTeams():
+    teams = db.session.query(Team.team_name).all() 
+    team_list = [team.team_name for team in teams]
+    return team_list
+
+
+def getPlayersByTeam(team_name):
```
Comment (Jarpletz) [inline Team_Roster/immaculateGridCalculations/gridSolver.py:17]: It looks to me like it is fully making the query instead of returning a subquery which will be ran later (i.e. as a different transaction). Is this inturpretation correct? Can we utilize subqueries as I believe we do elsewhere to only run a single transaction?

File: Team_Roster/immaculateGridCalculations/gridSolver.py [hunk 2]
```diff
@@ -1,6 +1,110 @@
...
+        .filter(Pitching.p_W >= numWins)
+        .group_by(People.playerID)
+    )
+    print(f"Players found with >= {numWins} wins: {playerWins.count()}")
+
+    return playerWins.all()
+
+
 
 def solveGrid(questions):
```
Comment (Jarpletz) [inline Team_Roster/immaculateGridCalculations/gridSolver.py:61]: It is not the clearest which part of this function is doing what. Given that we will have so many people touching it, I'd appreciate it if you add comments explaining what everything is doing, and where and how everyone should add their queries.

File: Team_Roster/immaculateGridCalculations/gridSolver.py [hunk 3]
```diff
@@ -1,6 +1,110 @@
...
+    seenPlayers = set()
+
+    teamList = getAllTeams()
+    print("Team List:", teamList)  # Debug team mapping
+
+    for currentQuestion in questions:
+        if currentQuestion in teamList:
+            queryAnswers.append(getPlayersByTeam(currentQuestion.strip()))
+
+        if "Win Season" in currentQuestion:
```
Comment (Jarpletz) [inline Team_Roster/immaculateGridCalculations/gridSolver.py]: *elif

File: Team_Roster/immaculateGridCalculations/gridSolver.py [hunk 4]
```diff
@@ -1,6 +1,110 @@
...
+    rowQueries = queryAnswers[3:]
+
+    finalPlayers = []
+
+    for colQuery in columnQueries:
+        columnPlayers = {player.playerID for player in colQuery}
+        for rowQuery in rowQueries:
+            rowPlayers = {player.playerID for player in rowQuery}
+            
+            matching = columnPlayers & rowPlayers
```
Comment (Jarpletz) [inline Team_Roster/immaculateGridCalculations/gridSolver.py]: again, more comments would be great here

**Review by Jarpletz — CHANGES_REQUESTED** [review]
For the most part, I'm happy with this, good work. However, since so many people will be touching this file, I think we would all greatly benefit from if this file was heavily commented - especially since we are not all sqAlchemy pros yet, and to establish a standard for others.

File: Team_Roster/immaculateGridCalculations/gridSolver.py (hunk 1 above)
Comment (Renzenar) [reply Team_Roster/immaculateGridCalculations/gridSolver.py:17]: What exactly do you mean by single transaction? Unless I misunderstand your question, what I see happening elsewhere is taking a single query and running it later. What is happening here is I am taking the results of the query and finding the common factors in each result to determine an answer.

**Comment by Renzenar** [comment]
Good point about the comments. I tried to make everything I was doing and thinking more clear.

File: Team_Roster/immaculateGridCalculations/gridSolver.py (hunk 1 above)
Comment (Jarpletz) [reply Team_Roster/immaculateGridCalculations/gridSolver.py:17]: My apologies let me rephrase- 
It appears to me right now that we are doing the join in memory- making two separate queries, fetching the results, then taking the common factors of the two arrays. I believe there might be a way in SQAlchemy to make a single query equivalent to
SELECT FROM x JOIN y LIMIT 1 
where x and y are subqueries returned by the other functions

**Review by Jarpletz — APPROVED** [review]
I think this is doing what we want it to now, great work! I think there may be a way to do the last part in one query (instead of a separate query to fetch the name), and to treat the subqueries as a join instead of a filter. However, I think this is good enough to get everyone else on it, so I’d say it’s good to go for now if you’re happy with it. Good work!
